File: 438.找到字符串中所有字母异位词.py
#
# @lc app=leetcode.cn id=438 lang=python3
#
# [438] 找到字符串中所有字母异位词
#
from collections import defaultdict
from typing import List

# @lc code=start
class Solution:
    # sliding window
    # T: O(len(s) + len(p))
    # S: O(alpha_size)
    def findAnagrams(self, s: str, p: str) -> List[int]:
        required_chars = defaultdict(int)
        for c in p:
            required_chars[c] += 1
        window = defaultdict(int)

        l = r = valid_cnt = 0
        res = []
        while r < len(s):
            # Expand window to the right
            c_r = s[r]
            r += 1
            if c_r in required_chars:
                window[c_r] += 1
                if window[c_r] == required_chars[c_r]:
                    valid_cnt += 1
            
            while r - l >= len(p):
                if valid_cnt == len(required_chars):
                    res.append(l)

                # Try shrinking window by moving the left pointer,
                # then update the counts.
                c_l = s[l]
                if c_l in required_chars:
                    if window[c_l] == required_chars[c_l]:
                        valid_cnt -= 1
                    window[c_l] -= 1
                l += 1
        return res
        
# @lc code=end

File: 438.找到字符串中所有字母异位词.java
/*
 * @lc app=leetcode.cn id=438 lang=java
 *
 * [438] 找到字符串中所有字母异位词
 */
import java.util.*;

// @lc code=start
class Solution {
    public List<Integer> findAnagrams(String s, String p) {
        int[] req = new int[26], win = new int[26];
        int reqChTypes = 0;
        for (int i = 0; i < p.length(); ++i) {
            int cIdx = (int) (p.charAt(i) - 'a');
            if (req[cIdx] == 0)  ++reqChTypes;
            ++req[cIdx];
        }

        int l = 0, r = 0, validChTypes = 0;
        List<Integer> res = new ArrayList<>();
        while (r < s.length()) {
            // Expand window to the right
            int cIdx = (int) (s.charAt(r) - 'a');
            ++r;
            if (req[cIdx] > 0) {
                ++win[cIdx];
                if (win[cIdx] == req[cIdx])  ++validChTypes;
            }

            while (r - l >= p.length()) {
                if (validChTypes == reqChTypes) {
                    res.add(l);
                }

                // Try shrinking window by moving the left pointer,
                // then update the counts.
                int cIdxD = (int) (s.charAt(l) - 'a');
                ++l;
                if (req[cIdxD] > 0) {
                    if (win[cIdxD] == req[cIdxD])  --validChTypes;
                    --win[cIdxD];
                }
            }
        }
        return res;
    }
}
// @lc code=end

